lib/msf/core/post/windows/version.rb

```diff
@@ -0,0 +1,150 @@
+# -*- coding: binary -*-
+
+module Msf::Post::Windows::Version
+include Msf::Post::Windows::Registry
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Compat' => {
+          'Meterpreter' => {
+            'Commands' => %w[
+              stdapi_railgun_api
+            ]
+          }
+        }
+      )
+    )
+  end
+
+  def get_version_info
+    result = get_version_info_impl
+    if result.nil?
+      print_error("Couldn't retrieve the target's build number!")
+      raise RuntimeError.new("Couldn't retrieve the target's build number!")
+    end
+
+    result
+  end
+
+  def get_version_info_fallback_impl
+        build_num_raw = cmd_exec('ver')
+        groups = build_num_raw.match(/.*Version\s+(\d+)\.(\d+)\.(\d+)(\.(\d+))?/)
+        if groups.nil?
+          return nil
+        end
+
+        major, minor, build, unused, revision = groups.captures
+        revision = 0 if revision.nil?
+        # Default to workstation, since it'll likely be an older OS - pre Server editions
+        return Msf::WindowsVersion.new(major.to_i, minor.to_i, build.to_i, 0, Msf::WindowsVersion::VER_NT_WORKSTATION)
+  end
+
+  def get_version_info_impl
+    if session.type == 'meterpreter'
+      result = session.railgun.ntdll.RtlGetVersion(input_os_version_info_ex)
+      os_version_info_ex = unpack_version_info(result['VersionInformation'])
+      major = os_version_info_ex[1]
+      minor = os_version_info_ex[2]
+      build = os_version_info_ex[3]
+      service_pack = os_version_info_ex[6]
+      product_type = os_version_info_ex[9]
+
+      Msf::WindowsVersion.new(major, minor, build, service_pack, product_type)
+    else
+      # Command shell - we'll try reg commands, and fall back to `ver`
+      build_str = shell_registry_getvaldata('HKLM\SOFTWARE\Microsoft\Windows NT\CurrentVersion', 'CurrentBuildNumber', Msf::Post::Windows::Registry::REGISTRY_VIEW_NATIVE)
+      if build_str.nil?
+        return get_version_info_fallback_impl
+      end
+
+      version_str = shell_registry_getvaldata('HKLM\SOFTWARE\Microsoft\Windows NT\CurrentVersion', 'CurrentVersion', Msf::Post::Windows::Registry::REGISTRY_VIEW_NATIVE)
+      if version_str.nil?
+        return get_version_info_fallback_impl
+      end
+
+      build_num = build_str.to_i
+      version_match = version_str.match(/(\d+)\.(\d+)/)
+      if version_match.nil?
+        return get_version_info_fallback_impl
+      end
+
+      major, minor = version_match.captures
+      major = major.to_i
+      minor = minor.to_i
+        
+      product = shell_registry_getvaldata('HKLM\SYSTEM\CurrentControlSet\Control\ProductOptions', 'ProductType', Msf::Post::Windows::Registry::REGISTRY_VIEW_NATIVE)
+      case product
+      when /WinNT/
+        product_type = Msf::WindowsVersion::VER_NT_WORKSTATION
+      when /LanmanNT/
+        product_type = Msf::WindowsVersion::VER_NT_DOMAIN_CONTROLLER
+      when /ServerNT/
```

Comment:
Do we need regex here? I'm under the impression this is a simple string comparison, which is usually preferred when not needed:
```suggestion
      when 'WinNT'
        product_type = Msf::WindowsVersion::VER_NT_WORKSTATION
      when 'LanmanNT'
        product_type = Msf::WindowsVersion::VER_NT_DOMAIN_CONTROLLER
      when 'ServerNT'
```
also, maybe using constants would make sense instead of raw strings here.

Reply:
Good call.